### ISSUE TITLE
`MPI_File_open` hangs every second time it is called

### ISSUE THREAD

**Issue by A-CGray** [open-mpi/ompi#11934]
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
v4.1.4/5

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
Installed from official release tarball

```shell
$ echo $MPI_DIR
/home/ali/packages/openmpi-4.1.5/opt-gfortran
$ ./configure -prefix=$MPI_DIR
$ make -j14 all install
```

### Please describe the system on which you are running

* Operating system/version: Manjaro KDE (Linux kernel 6.1.53-1)
* Computer hardware: AMD Ryzen 9 5950X CPU, MSI MAG X570 TOMAHAWK WIFI motherboard
* Network type:

-----------------------------

## Details of the problem

As the title suggests, I have been experiencing an issue for the past several weeks in some codes I develop where `MPI_File_open` hangs every second time it is called (e.g I run a program that calls `MPI_File_open` once and it hangs, next time I run it doesn't, next time I run it does, etc ...). 

Here is a minimal working example that displays this behaviour:

```c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char* argv[]) {
  MPI_Init(&argc, &argv);

  int my_rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

  MPI_File handle;
  int access_mode = MPI_MODE_CREATE /* Create the file if it does not exist */
                    | MPI_MODE_WRONLY; /* Open the file for writing only */

  printf("[MPI process %02d] About to open file.\n", my_rank);
  if (MPI_File_open(MPI_COMM_WORLD, "file.tmp", access_mode, MPI_INFO_NULL,
                    &handle) != MPI_SUCCESS) {
    printf("[MPI process %02d] Failure in opening the file.\n", my_rank);
    MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
  }

  printf("[MPI process %02d] File opened successfully.\n", my_rank);

  if (MPI_File_close(&handle) != MPI_SUCCESS) {
    printf("[MPI process %02d] Failure in closing the file.\n", my_rank);
    MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
  }

  printf("[MPI process %02d] File closed successfully.\n", my_rank);

  MPI_Finalize();

  return EXIT_SUCCESS;
}
```

And an example of the behaviour:
```shell
$ mpicc MPIWriteTest.c
$ mpiexec -n 2 ./a.out 
[MPI process 00] About to open file.
[MPI process 01] About to open file.
[MPI process 00] File opened successfully.
[MPI process 01] File opened successfully.
[MPI process 00] File closed successfully.
[MPI process 01] File closed successfully.

$ mpiexec -n 2 ./a.out
[MPI process 00] About to open file.
[MPI process 01] About to open file.
^C% <<< me killing the program because it is hanging

$ mpiexec -n 2 ./a.out
[MPI process 00] About to open file.
[MPI process 01] About to open file.
[MPI process 01] File opened successfully.
[MPI process 00] File opened successfully.
[MPI process 01] File closed successfully.
[MPI process 00] File closed successfully.

$ mpiexec -n 2 ./a.out
[MPI process 00] About to open file.
[MPI process 01] About to open file.
^C%
```

This behaviour occurs whether or not the file being opened already exists or not.

This behaviour does not coincide with any changes in the version of OpenMPI I am running. I first encountered it around 2023-08-21 on v4.1.4, which I had been running without issue for almost a year at that point. I have since upgraded to v4.1.5 but this didn’t change anything. This makes me think that the issue is due to something that occurred during a Manjaro update.

My first suspicion was that this was caused by the upgrade to gcc-13 in the [2023-06-04 stable update](https://forum.manjaro.org/t/stable-update-2023-06-04-kernels-gnome-44-1-plasma-5-27-5-python-3-11-toolchain-firefox/141610). I therefore tried recompiling OpenMPI using gcc-12, but this had no effect.

Despite the fact that I think this issue has been caused by a Manjaro OS update, I'm opening this issue to see if you have any ideas of what the cause may be?

**Comment by loqs** [comment]
I realized this issue may not be the same as #11913 which was introduced by a cherry-pick of https://github.com/open-mpi/ompi/commit/6e8e14f2c2f207d5fa51299cc67558697a5b7d63 used in Arch Linux in 4.1.5-3 and 4.1.5-4 is also present in 4.1.6 but not 4.1.5 or 4.1.4.

**Comment by jsquyres** [comment]
@loqs Your initial description says that you are installing from an Open MPI release tarball, but later you seem to mention patches from Arch Linux.  Can you clarify exactly what source you are building?

**Comment by A-CGray** [comment]
Just discovered that this bug doesn't occur if I use the system installed ompi package (@christian-heusel 's openmpi 4.1.5-5). Only if I use the versions (4.1.4/4.1.5) that I built myself from the release tarballs

**Comment by jsquyres** [comment]
@A-CGray I'm still trying to understand.  Is the current status:

1. The un-patched Open MPI v4.1.4 release tarball *does* contain the issue
1. The un-patched Open MPI v4.1.5 release tarball *does* contain the issue
1. The Arch system-installed Open MPI v4.1.5-5 has some magic that seems to resolve the issue

If these statements are correct, then I'm a bit confused about the statements regarding 6e8e14f2c2f207d5fa51299cc67558697a5b7d63 because the v1.4.5 release tarball contains that commit.

Can you clarify?

**Comment by A-CGray** [comment]
> @A-CGray I'm still trying to understand. Is the current status:
> 
>     1. The un-patched Open MPI v4.1.4 release tarball _does_ contain the issue
> 
>     2. The un-patched Open MPI v4.1.5 release tarball _does_ contain the issue
> 
>     3. The Arch system-installed Open MPI v4.1.5-5 has some magic that seems to resolve the issue
> 
> 
> If these statements are correct, then I'm a bit confused about the statements regarding [6e8e14f](https://github.com/open-mpi/ompi/commit/6e8e14f2c2f207d5fa51299cc67558697a5b7d63) because the v1.4.5 release tarball contains that commit.
> 
> Can you clarify?

Yes your statements are correct, tarball builds hang, [arch linux package](https://archlinux.org/packages/extra/x86_64/openmpi/) does not.

I'm not familiar with https://github.com/open-mpi/ompi/commit/6e8e14f2c2f207d5fa51299cc67558697a5b7d63 other than what @loqs said in their comments.

**Comment by loqs** [comment]
> If these statements are correct, then I'm a bit confused about the statements regarding https://github.com/open-mpi/ompi/commit/6e8e14f2c2f207d5fa51299cc67558697a5b7d63 because the v1.4.5 release tarball contains that commit.

https://github.com/open-mpi/ompi/commit/6e8e14f2c2f207d5fa51299cc67558697a5b7d63 is in the v1.4.5 tarball while not being contained in the 4.1.5 tag?
```
 git tag --contains 6e8e14f2c2f207d5fa51299cc67558697a5b7d63
v4.1.6rc1
v4.1.6rc2
```

openmpi 4.1.5-5 removed the patches previously applied and is building from the tarball [1][2].

[1] https://gitlab.archlinux.org/archlinux/packaging/packages/openmpi/-/commit/fea2f1ad4970cfcebd9b674943f7e129f4b1837d
[2] https://gitlab.archlinux.org/archlinux/packaging/packages/openmpi/-/blob/fea2f1ad4970cfcebd9b674943f7e129f4b1837d/PKGBUILD

**Comment by loqs** [comment]
@A-CGray does the older Arch release 4.1.5-4 produce the issue?
I suspect as I wrote in https://github.com/open-mpi/ompi/issues/11934#issuecomment-1727676382 this issue has similar symptoms but a different cause.

**Comment by jsquyres** [comment]
> https://github.com/open-mpi/ompi/commit/6e8e14f2c2f207d5fa51299cc67558697a5b7d63 is in the v1.4.5 tarball while not being contained in the 4.1.5 tag?

Sigh.  This is apparently me not being able to read properly.  You are absolutely correct -- 6e8e14f2c2f207d5fa51299cc67558697a5b7d63 is *not* in v4.1.5 -- it *is* in the upcoming 4.1.6.

Could you see if the latest 4.1.6rc tarball works for you?  https://www.open-mpi.org/software/ompi/v4.1/

**Comment by loqs** [comment]
Both the reproducer from this issue and the reproducer from https://github.com/open-mpi/ompi/issues/11913 fail with 4.1.6rc2 same result with the v4.1.x branch HEAD https://github.com/open-mpi/ompi/commit/d06bc98eac1cab813ea192b21493c9a2e0c46d96

**Comment by jsquyres** [comment]
> Both the reproducer from this issue and the reproducer from #11913 fail with 4.1.6rc2 same result with the v4.1.x branch HEAD [d06bc98](https://github.com/open-mpi/ompi/commit/d06bc98eac1cab813ea192b21493c9a2e0c46d96)

That would seem to imply that https://github.com/open-mpi/ompi/commit/6e8e14f2c2f207d5fa51299cc67558697a5b7d63 is not the fix, and that the fix is something else in the Arch-specific patches for v4.1.5-5.

**Comment by jsquyres** [comment]
Wait, if 6e8e14f is the *cause* (and not the solution), then I'm confused -- @A-CGray said these statements were correct:

1. The un-patched Open MPI v4.1.4 release tarball _does_ contain the issue
3. The un-patched Open MPI v4.1.5 release tarball _does_ contain the issue
4. The Arch system-installed Open MPI v4.1.5-5 has some magic that seems to resolve the issue

The un-patched 4.1.4 and 4.1.5 release tarballs do not contain 6e8e14f.  And you're saying that the Arch system-installed Open MPI v4.1.5-5 had all patched removed, so it should be the same as the un-patched Open MPI v4.1.5 release tarball.

These now appear to be contradictory statements.  Can you clarify?

**Comment by jsquyres** [comment]
6e8e14f is about PMIx, which is external to Open MPI (but is bundled in Open MPI, if you don't have an external PMIx installed).

@edgargabriel Is PMIx involved in `MPI_File_open()` (in v4.1.x)?

**Comment by jsquyres** [comment]
Thanks @edgargabriel.  Assumedly you were using the internal PMIx, right?

@loqs What (external) version of PMIx are you using?

**Comment by edgargabriel** [comment]
> Thanks @edgargabriel. Assumedly you were using the internal PMIx, right?

I assume so, but with the 4.1.x I usually didn't explicitly specify that

**Comment by loqs** [comment]
openpmix 4.2.6 no patches configured with ```./configure --prefix=/usr --sysconfdir=/etc/openpmix```

**Comment by A-CGray** [comment]
I think @loqs is correct, configuring my tarball build with `--with-pmix=external` fixes this issue

**Comment by rhc54** [comment]
I think I am totally confused 😄 

PMIx has nothing to do with MPI_File_open (or any other MPI_File operation). The commit you are citing was done solely to deal with _external_ PMIx installations - OMPI needed to use the proper macros for forward compatibility with PMIx versions. It should have no impact when working with the _internal_ PMIx installation.

I'd be suspicious that there is something else wrong, and switching between PMIx versions is only masking the problem in some situations. I'm always leery when someone says that a distro "patched" an official release - that almost never works out well.

**Comment by loqs** [comment]
@A-CGray does the issue come back if you use the 4.1.5rc2 tarball still with ```--with-pmix=external```

**Comment by jsquyres** [comment]
We have a few v4.1.x patches coming; hopefully they will fix the issue.  Stand by...

**Comment by A-CGray** [comment]
Great, thanks to you and @loqs for all your help in this thread. I do at least have a working build now (either using the Arch package or configuringn the tarball with external pmix) even if we don't know what the underlying cause is. Let me know if there are any other builds you want me to test out.

**Comment by jsquyres** [comment]
@A-CGray Could you try the HEAD of the `v4.1.x` branch?  Or you can wait for a nightly snapshot tarball to be created this evening (https://www.open-mpi.org/nightly/v4.1.x/).

**Comment by loqs** [comment]
The issue is still present in v4.1.x as of 1bafe60d07effda8479cebbe47d7988b85784638.

**Comment by loqs** [comment]
1bafe60d07effda8479cebbe47d7988b85784638 with 12025000b00eed3bef5a9f907cdfab5b8f563010 and 6e8e14f2c2f207d5fa51299cc67558697a5b7d63 reverted no longer produce the issue.  Reverting just 12025000b00eed3bef5a9f907cdfab5b8f563010 still produces the issue.  6e8e14f2c2f207d5fa51299cc67558697a5b7d63 does not revert cleanly without first reverting 12025000b00eed3bef5a9f907cdfab5b8f563010.

**Comment by ggouaillardet** [comment]
TL;DR looks like a `glibc` issue, both Open MPI and PMIx are fine.

I was able to reproduce the issue, but only on an up-to-date archlinux.

in `ompi/mca/sharedfp/sm/sharedfp_sm_file_open.c`, both processes `sem_open(OMPIO_file.tmp, O_CREAT)`

And every two executions, one fail (!)

Here is the `strace` output on the successful rank
```
write(2, "trying to sem_open(OMPIO_file.tm"..., 35) = 35
openat(AT_FDCWD, "/dev/shm/sem.OMPIO_file.tmp", O_RDWR|O_NOFOLLOW|O_CLOEXEC) = -1 ENOENT (No such file or directory)
getrandom("\x74\x63\x64\x19\x5f\xcc\x58\x3a", 8, GRND_NONBLOCK) = 8
newfstatat(AT_FDCWD, "/dev/shm/sem.8Y5JJR", 0x7ffc2a13dfd0, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/dev/shm/sem.8Y5JJR", O_RDWR|O_CREAT|O_EXCL|O_CLOEXEC, 0644) = 17
write(17, "\1\0\0\0\0\0\0\0\200\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 32) = 32
mmap(NULL, 32, PROT_READ|PROT_WRITE, MAP_SHARED, 17, 0) = 0x7f5b64661000
link("/dev/shm/sem.8Y5JJR", "/dev/shm/sem.OMPIO_file.tmp") = 0
newfstatat(17, "", {st_mode=S_IFREG|0644, st_size=32, ...}, AT_EMPTY_PATH) = 0
unlink("/dev/shm/sem.8Y5JJR")           = 0
close(17)                               = 0
write(2, "success\n", 8)                = 8
```

and here is the `strace` output on the rank that fails
```
write(2, "trying to sem_open(OMPIO_file.tm"..., 35) = 35
openat(AT_FDCWD, "/dev/shm/sem.OMPIO_file.tmp", O_RDWR|O_NOFOLLOW|O_CLOEXEC) = -1 ENOENT (No such file or directory)
getrandom("\x3b\x1d\x9c\x13\x34\xc1\xdb\x42", 8, GRND_NONBLOCK) = 8
newfstatat(AT_FDCWD, "/dev/shm/sem.9ljA8e", 0x7fff50e3b7d0, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/dev/shm/sem.9ljA8e", O_RDWR|O_CREAT|O_EXCL|O_CLOEXEC, 0644) = 17
write(17, "\1\0\0\0\0\0\0\0\200\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 32) = 32
mmap(NULL, 32, PROT_READ|PROT_WRITE, MAP_SHARED, 17, 0) = 0x7f9383842000
link("/dev/shm/sem.9ljA8e", "/dev/shm/sem.OMPIO_file.tmp") = -1 EEXIST (File exists)
munmap(0x7f9383842000, 32)              = 0
unlink("/dev/shm/sem.9ljA8e")           = 0
close(17)                               = 0
openat(AT_FDCWD, "/dev/shm/sem.OMPIO_file.tmp", O_RDWR|O_CREAT|O_EXCL|O_CLOEXEC, 000) = -1 EEXIST (File exists)
write(2, "failure: File exists\n", 21)  = 21
```

So `sem_open()` basically does this
- open the semaphore file
- if the file does not exist open a random file name and initialize it and then hard link it to the original file

when a race condition occurs, `link()` fails but for some reason, `sem_open()` returns instead of simply restarting from the beginning. (And this is what `glibc-2.17.326.el7_9` is doing)

At first glance, that looks like a bug in the `glibc`.

Note I just upgraded to the latest glibc `2.38-4` and previously ran `2.37.3`.
If one of you have a previous `glibc` running, it would be worth trying it.

I guess that reverting some commits re-introduce a bug and Open MPI believes ranks are running on different nodes, and hence `sharedfp/sm` is not used and that is enough to "hide" the bug.


For the time being, a simple workaround is to avoid using `sem_open()`:
```
ac_cv_func_sem_open=no configure ...
```

**Comment by ggouaillardet** [comment]
About the glibc bug, there is indeed a retry, but it seems `sem_open()` returns `-1` with `errno=EEXIST` anyway (!).

**Comment by loqs** [comment]
Adding ac_cv_func_sem_open=no does workaround the issue for me.  Are there any negative consequences of this workaround if Arch was to incoporate it until the issue in glibc is fixed?

**Comment by ggouaillardet** [comment]
Open MPI will use `sem_init()` instead of `sem_open()`, so it should be fine.

If I may, I'd rather recommend you fix the glibc (a fix has been suggested, or you can simply revert the changes related to `sem_open.c`)

**Comment by loqs** [comment]
@freswa would are your thoughts?

**Comment by freswa** [comment]
Sounds like a reasonable explanation for the bug. I'd like to wait with a fix until it has been reviewed and commited.

**Comment by jsquyres** [comment]
@ggouaillardet Wow -- great job finding all of this!

**Comment by A-CGray** [comment]
Yeah great work @ggouaillardet , this would also explain why the onset of this error on my machine coincided with an OS update and not any changes I made to my OpenMPI installation.

Can confirm that this configuration with the 4.1.5 tarball gives me a working build:

```shell
./configure --prefix $MPI_DIR \
ac_cv_func_sem_open=no \ 
--enable-pretty-print-stacktrace \
--with-cuda=/opt/cuda \
--with-valgrind
```

**Comment by jsquyres** [comment]
Based on the comments here, I think the conclusion is that there's a bug in glibc, and Open MPI is fine.  There is a workaround if someone has this buggy version of glibc, so I don't think there's anything Open MPI needs to do.

I'm going to close this issue.
